File: answers/other/coding/Linked List Cycle.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        
        if None == head:
        	return False
        left = head
        right = head.next
        while right and left != right:
        	left = left.next
        	right = right.next
        	if not right:
        		return False
        	right = right.next
        return None != right

File: answers/other/answers/lintcode/Linked List Cycle.cpp
/**
 * Definition of ListNode
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 *     ListNode(int val) {
 *         this->val = val;
 *         this->next = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param head: The first node of linked list.
     * @return: True if it has a cycle, or false
     */
    bool hasCycle(ListNode *head) {
        // write your code here
        if (NULL == head) {
            return false;
        }
        ListNode *left = head;
        ListNode *right = head->next;
        while (left && right && right->next) {
            if (left == right) {
                return true;
            }
            left = left->next;
            right = right->next->next;
        }
        return false;
    }
};



